planner/core/logical_plan_builder.go

```diff
@@ -3734,13 +3734,38 @@ func checkUpdateList(ctx sessionctx.Context, tblID2table map[int64]table.Table,
 	if err != nil {
 		return err
 	}
+	isPKUpdated := make(map[int64]model.CIStr)
```

Comment:
The newly added logic makes this function difficult to understand.
I bet I can't understand it after 3 months.

From the `updt.TblColPosInfos`, we have the tables and columns to be updated.

```
t1          c2 c4 c6 ...
t2           c1  c2 c5 c7 ...
...
```

and what we want to prevent is this case:
t1 and t2 are actually the joinings of the same table `t`,
and they both want to update the `c2` column, which is a primary key.

So, how about handling it like this?
Step1: find out the common columns  in `the updt.TblColPosInfos`
       This can be done by sort using `table id + column id` and deduplicate
Step2: is the common column the primary key?
      If so, return the error

Reply:
Sort + deduplicate may not be suitable for primary key of multiple columns. As shown in the following cases:
``` mysql
create table t (a int, b int, c int, primary key(a,c));
update t set t.a=10, t.c=20; # reporting error
update t m, t n set m.a=10, m.b=20, n.c=30; # not reporting error
```
It would be more complicated to distinguish primary keys from different table alias.